go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,359 @@
+		if err != nil {
+			return err
+		}
+		// expect an empty account
+		if err := verifyAccountProof(rootHash, proof, addr, AccountInfo{}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// verifyProofEmptyStorage verifies the consistency of witness proofs for empty storage that are not present in the trie.
+func verifyProofEmptyStorage(trie *LiveTrie, rootHash common.Hash, addr common.Address) error {
+	const numKeys = 10
+	keys, err := generateUnknownKeys(trie, numKeys, addr)
+	if err != nil {
+		return err
+	}
+	proof, err := trie.CreateWitnessProof(addr, keys...)
+	if err != nil {
+		return err
+	}
+	if err := verifyStorageProof(rootHash, proof, addr, keys, nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// verifyAccountProof verifies the consistency between the input witness proofs and the account.
+func verifyAccountProof(root common.Hash, proof witness.Proof, addr common.Address, info AccountInfo) error {
+	balance, complete, err := proof.GetBalance(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := balance, info.Balance; got != want {
+		return fmt.Errorf("balance mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+	nonce, complete, err := proof.GetNonce(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := nonce, info.Nonce; got != want {
+		return fmt.Errorf("nonce mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+	code, complete, err := proof.GetCodeHash(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := code, info.CodeHash; got != want {
+		return fmt.Errorf("code mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+
+	return nil
+}
+
+// verifyStorageProof verifies the consistency between the input witness proofs and the storage.
+func verifyStorageProof(root common.Hash, proof witness.Proof, addr common.Address, keys []common.Key, storage map[common.Key]common.Value) error {
+	for _, key := range keys {
+		proofValue, complete, err := proof.GetState(root, addr, key)
+		if err != nil {
+			return err
+		}
+		if !complete {
+			return fmt.Errorf("proof incomplete for address: 0x%x, key: 0x%x", addr, key)
+		}
+		if got, want := proofValue, storage[key]; got != want {
+			return fmt.Errorf("storage mismatch for 0x%x, key 0x%x, got %v, want %v", addr, key, got, want)
+		}
+	}
+
+	return nil
+}
+
+// generateUnknownAddresses generates a slice of addresses that does not appear in the input MPT.
+// if in an unlikely situation a generated address is in the trie, it is not added in the result list.
+// It means that in an unlikely situation, the result list may contain fewer addresses than the input number.
+func generateUnknownAddresses(trie *LiveTrie, number int) ([]common.Address, error) {
```

Comment:
nit: rename to "generateUnusedAddresses"

Reply:
ok, done